feature/recoverable_signatures

regcostajr opened the pull request:
Implements https://github.com/Perl-Bitcoin/Bitcoin-Secp256k1/issues/3

Adds recoverable signatures support:

New methods:

`sign_digest_recoverable`
~~`serialize_compact_recoverable`~~
`recover_public_key`
`recover_public_key_message`
`verify_message_recoverable`
`verify_digest_recoverable`

Usage:
```perl
my $recoverable_signature = $secp256k1->sign_message_recoverable($private_key, $message);

# (removed) my $compact_data = $secp256k1->serialize_compact_recoverable($recoverable_signature);

my $signature = $recoverable_signature->{signature};
my $recover_id = $recoverable_signature->{recovery_id};

my $recovered_pubkey = $secp256k1->recover_public_key_message($recoverable_signature, $message);
my $valid = $secp256k1->verify_message_recoverable($public_key, $recoverable_signature, $message);
```

This PR depends on https://github.com/Perl-Bitcoin/Alien-libsecp256k1/pull/1

regcostajr committed:
feature/recoverable_signatures

bbrtj commented on Secp256k1.xs:
```diff
@@ -626,6 +675,117 @@ _sign_schnorr(self, privkey, message)
+void
+_recover_pubkey_recoverable(self, recoverable_signature, message_hash)
+		SV* self
+		SV* recoverable_signature
```
This misses the point of low / high level API distinction. The signature should be passed in and out of XS through `_signature_recoverable` once you have implemented it. Having it as an argument here is a duplication in the low level API

regcostajr replied:
I've changed it to use the signature over `_signature_recoverable`, I was trying to make `recover_pubkey` to not depend on it, but it does makes sense this way

bbrtj commented on t/api.t:
```diff
@@ -117,5 +117,90 @@ subtest 'should combine' => sub {
+		'wrong digest verification fails ok';
+};
+
+subtest 'ethereum yellowpaper specification' => sub {
```
I would appreciate a link to the specification here (as a comment)

regcostajr replied:
done

bbrtj reviewed (changes requested):
I'm trying to wrap my head around these recoverable signatures. There seems to be a rather large flaw in the coding - we should not memcpy 65 bytes in and out of `secp256k1_ecdsa_recoverable_signature` structure. Schnorr signatures use this kind of memcpy because they have no special data structure declared for them. For recoverable signatures, I see there is a `secp256k1_ecdsa_recoverable_signature_parse_compact` and `secp256k1_ecdsa_recoverable_signature_serialize_compact` functions and those are what we should probably use. Comments say:

```
 *  The exact representation of data inside is implementation defined and not
 *  guaranteed to be portable between different platforms or versions. It is
 *  however guaranteed to be 65 bytes in size, and can be safely copied/moved.
 *  If you need to convert to a format suitable for storage or transmission, use
 *  the secp256k1_ecdsa_signature_serialize_* and
 *  secp256k1_ecdsa_signature_parse_* functions.
```

I read this as: you can move 65 bytes from one structure to the other, otherwise use these methods. I would suggest:
- replace all instances of manual moving data into the structure with `parse_compact` method
- replace all instances of manual moving data out of the structure with `serialize_compact` method
- since these methods require 64 bytes + integer recovery id, find out if it is better in the API to use 65 bytes bytestring for it, or 64 bytes + an integer

If there is a spec for it, does it define how the recoverable signatures are serialized? Maybe the 65 byte representation is more custom than 64 bytes + an integer separately? I searched for some spec but it's not easy to find - would be grateful if you could share some resources.

regcostajr committed:
review - remove memcpy & follow existing impl

regcostajr commented:
@bbrtj thanks, I was over complicating this, I made the changes as per your review

```
- Removed raw memcpy
- Now using secp256k1_ecdsa_recoverable_signature_serialize_compact() to extract signature data
- Now using secp256k1_ecdsa_recoverable_signature_parse_compact() to create signatures from data
- API format: 64 bytes + recovery_id
- Changed from 65-byte blob to { signature => 64_bytes, recovery_id => 0..3 } hash refs
- _sign_recoverable() now stores in object state (like _sign())
- Fixed _signature_recoverable() getter/setter (now like _signature())
- _recover_pubkey_recoverable() works on stored signature (like _verify())
```

regcostajr commented on Secp256k1.xs:
```diff
@@ -18,12 +20,14 @@ typedef struct {
 	unsigned int pubkeys_count;
 	secp256k1_ecdsa_signature *signature;
 	unsigned char *schnorr_signature;
+    secp256k1_ecdsa_recoverable_signature *recoverable_signature;
```
done

regcostajr committed:
use constant instead of hardcode

bbrtj reviewed (changes requested):
Thanks, this looks much better now. I think both high and low level APIs are pretty much done, but I don't like the current internal representation of recoverable signature in `secp256k1_perl` structure. I think it should hold `secp256k1_ecdsa_recoverable_signature` structure (as it did previously) instead of signature and recovery_id separately, but the low level API should parse it on input and serialize it on output. This won't change how this feature works, only changes the internal structure and when parsing/serialization is done - only `_signature_recoverable` will call parse/serialize functions. 

The hashref structure of recoverable signature works well, may as well be a 2-element arrayref, but it is good as it is.

For other nitpicks:
- high level API's `recover_pubic_key` should be renamed to `recover_public_key_digest` so that the naming is consistent with other methods (having either `message` or `digest` in their names)
- new code does not seem to leak, but memory leak test was not updated. Here is a diff for my very naive test for leaks: https://irc.bbrtj.eu/file/1/JgMmxMcHCkuyRbxs.diff
- I released Alien::libsecp256k1 with the compilation flag for recoverable signatures, so please include that version as a dependency in `cpanfile`
- please update a changelog under `{{$NEXT}}`

regcostajr committed:
reviews - use secp256k1_ecdsa_recoverable_signature

regcostajr committed:
adding Changes

bbrtj reviewed (approved):
Alien::libsecp256k1 dependency version is not covered. But that's okay, I can change it myself before releasing. Rest of the code looks great now, thanks.
